timemachine/cpp/src/device_buffer.cu

```diff
@@ -4,34 +4,60 @@
 
 namespace timemachine {
 
-template <typename T> T *allocate(const std::size_t length) {
+template <typename T, bool PINNED> T *allocate(const std::size_t length) {
     if (length < 1) {
         throw std::runtime_error("device buffer length must at least be 1");
     }
     T *buffer;
-    gpuErrchk(cudaMalloc(&buffer, length * sizeof(T)));
+    if (PINNED) {
+        gpuErrchk(cudaMallocHost(&buffer, length * sizeof(T)));
+    } else {
+        gpuErrchk(cudaMalloc(&buffer, length * sizeof(T)));
+    }
     return buffer;
 }
 
-template <typename T>
-DeviceBuffer<T>::DeviceBuffer(const std::size_t length) : size(length * sizeof(T)), data(allocate<T>(length)) {}
+template <typename T, bool PINNED>
```

Comment:
It seems like it might be cleaner to introduce a separate class for pinned buffers, since the implementations seem almost entirely disjoint?

Reply:
Will do